modules/core/src/directx.cpp

```diff
@@ -1392,11 +1393,197 @@ void convertFromD3D11Texture2D(ID3D11Texture2D* pD3D11Texture2D, OutputArray dst
+    }
+
+    cl_device_id cl_device = static_cast<cl_device_id>(device.ptr());
+    size_t out_size = 0; // size in bytes
+    cl_int status = clGetDeviceInfo(cl_device, CL_DEVICE_PLATFORM, 0, nullptr, &out_size);
+    if (status != CL_SUCCESS) {
+        CV_Error(status, "OpenCL: clGetDeviceInfo failed");
+    }
+    cl_platform_id platform = nullptr; // The platform associated with this device.
+    status = clGetDeviceInfo(cl_device, CL_DEVICE_PLATFORM, out_size, &platform, nullptr);
+    if (status != CL_SUCCESS) {
+        CV_Error(status, "OpenCL: clGetDeviceInfo failed");
+    }
+    if (platform == nullptr) {
+        CV_Error(cv::Error::StsNullPtr, "OpenCL: clGetDeviceInfo returns empty platform");
+    }
+    auto&& platformName = PlatformInfo(&platform).name();
+
+    ID3D11Device *pD3D11Device = nullptr;
+    pD3D11Texture2D->GetDevice(&pD3D11Device);
+
+    if (pD3D11Device == nullptr) {
+        CV_Error(cv::Error::StsNullPtr, "OpenCL: GetDevice returns empty pD3D11Device in convertFromD3D11Texture2DtoCLMem");
+    }
+
+    if (!device.isExtensionSupported("cl_khr_d3d11_sharing")) {
+        CV_Error(cv::Error::OpenCLApiCallError, "OpenCL: Device doesn't support cl_khr_d3d11_sharing extension");
+    }
+
+    cl_context context = NULL;
+    cl_context_properties properties[] = {
+            CL_CONTEXT_PLATFORM, (cl_context_properties)platform,
+            CL_CONTEXT_D3D11_DEVICE_KHR, (cl_context_properties)(pD3D11Device),
+            CL_CONTEXT_INTEROP_USER_SYNC, CL_FALSE,
+            NULL, NULL
+    };
+    context = clCreateContext(properties, 1, &cl_device, NULL, NULL, &status);
+    if (status != CL_SUCCESS) {
+        CV_Error(status, "OpenCL: clCreateContext failed");
+    }
+
+    OpenCLExecutionContext clExecCtx;
+    try {
+        clExecCtx = OpenCLExecutionContext::create(platformName, platform, context, cl_device);
+        clExecCtx.getContext().setUserContext(std::make_shared<OpenCL_D3D11>(platform, pD3D11Device));
+    }
+    catch (...) {
+        if (context != nullptr) clReleaseContext(context);
+        throw;
+    }
+    clExecCtx.bind();
+    auto ctx = const_cast<Context&>(clExecCtx.getContext());
+
+    OpenCL_D3D11* impl = ctx.getUserContext<OpenCL_D3D11>().get();
+    if (nullptr == impl) {
+        CV_Error(cv::Error::OpenCLApiCallError, "OpenCL: Context initialized without DirectX interoperability");
+    }
+
+    cl_mem image2d = impl->clCreateFromD3D11Texture2DKHR(static_cast<cl_context>(ctx.ptr()),
+                                                         CL_MEM_READ_WRITE,
+                                                         pD3D11Texture2D,
+                                                         0,
+                                                         &status);
+    if (status != CL_SUCCESS) {
+        CV_Error(status, "OpenCL: clCreateFromD3D11Texture2DKHR failed");
+    }
+    auto image = Image2D::fromHandle(image2d);
+    clRetainMemObject(image2d);
```

Comment:
1. why do you use retainMemObject here since Image2D has taken ownership?
2. Since Image::fromHandle has added into public API and allow to it is sophisticated question "does it need OWNs handle or SHARE it inside  - who's responsible to delete image2d in this case"

3. I've checked  openCV code and found out that `from_handle ` is used inside class constructors as helper and also sometimes openCV called clReleaseHandle after constructor had finished in main creator function (I'm sorry, it's my fault to suggest this function name)


So, i propose: to do not interfere with openCV user experinece in using `from_handle` for Image2D ( in similar way as other from_handle) I suggest to rename it to `acquire_from_handle` (or something like that) which means:

a) private ctor for Image2D with cl_mem + boolean flag `owns` 
b) `acquire_from_handle`  invoked Image2D ctor with boolean flag = true
c) NO for clRetainHandle for cl_mem/cl_image in Image2D ctor for boolean flag == true ( and YES for owns == false)
d)  YES for clReleaseHandle  in Image2D dtor
e) NO for clRealeaseHandle for image2d in main after function  `acquire_from_handle` was called

@alalek @mpashchenkov  what do you think about such ideas?